File: Dynamic-Programming/560. Subarray Sum Equals K/subarraySumEqualsK.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
 var subarraySum = function(nums, k) {
    // sum[1:3] = sum[0:3] - sum[0:0] (inclusive)
    let map = new Map();
    // base case when sum is 0, it appears once
    // take into consideration of 0 index
    map.set(0, 1);
    let sum = 0, result = 0;
    
    for (let i = 0; i < nums.length; i++) {
        sum += nums[i];
        if (map.has(sum - k)) {
            result += map.get(sum - k);
        } 
        map.set(sum, (map.get(sum) || 0) + 1);
    }
    
    return result;
};

File: Dynamic-Programming/560. Subarray Sum Equals K/subarraySumEqualsK.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
    # k = sum k = 2
    # 1, 2, 3
    # if 3 - 2 in map, meaning between 3 and 1, there is a k
    
        result = 0
        sum = 0
        map = {}
        map[0] = 1
        
        for num in nums:
            sum += num
            
            if (sum - k) in map:
                result += map[sum - k]
                
            if sum in map:
                map[sum] += 1
            else:
                map[sum] = 1
        
        
        return result